cmake/Sanitizers.cmake

```diff
@@ -0,0 +1,70 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2013-2022 The Foundry Visionmongers Ltd
+
+function(enable_sanitizers project_name)
+
+    if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" OR CMAKE_CXX_COMPILER_ID MATCHES ".*Clang")
+        option(ENABLE_COVERAGE "Enable coverage reporting for gcc/clang" OFF)
+
+        if (ENABLE_COVERAGE)
+            target_compile_options(${project_name} PRIVATE --coverage -O0 -g)
+            target_link_libraries(${project_name} PRIVATE --coverage)
+        endif ()
+
+        set(SANITIZERS "")
+
+        option(ENABLE_SANITIZER_ADDRESS "Enable address sanitizer" OFF)
+        if (ENABLE_SANITIZER_ADDRESS)
+            list(APPEND SANITIZERS "address")
+        endif ()
+
+        option(ENABLE_SANITIZER_LEAK "Enable leak sanitizer" OFF)
+        if (ENABLE_SANITIZER_LEAK)
+            list(APPEND SANITIZERS "leak")
+        endif ()
+
+        option(ENABLE_SANITIZER_UNDEFINED_BEHAVIOR "Enable undefined behavior sanitizer" OFF)
+        if (ENABLE_SANITIZER_UNDEFINED_BEHAVIOR)
+            list(APPEND SANITIZERS "undefined")
+        endif ()
+
+        option(ENABLE_SANITIZER_THREAD "Enable thread sanitizer" OFF)
+        if (ENABLE_SANITIZER_THREAD)
+            if ("address" IN_LIST SANITIZERS OR "leak" IN_LIST SANITIZERS)
+                message(WARNING "Thread sanitizer does not work with Address and Leak sanitizer enabled")
```

Comment:
Is this warning based on experimentation?
Or is this a copy-n-paste from elsewhere?

I'm just surprised as to the need, and if it is a copy-n-paste, then there's a lot of duplication around.

Reply:
Its a copy-paste, but I have also experienced this in practice. Not sure what duplication you're seeing?

Reply:
Just checked and indeed the compile complains, interestingly at CMake configure time, but the error output is rather dense(!):
```
Cannot get compiler information:
	Compiler exited with error code 1: /usr/bin/g++-6 -xc++ -D_GLIBCXX_USE_CXX11_ABI=0 -Dopenassetio_EXPORTS -g -flto -fno-fat-lto-objects -fvisibility=hidden -fvisibility-inlines-hidden -Wall -Wextra -Wshadow -Wnon-virtual-dtor -Wold-style-cast -Wcast-align -Wunused -Woverloaded-virtual -Wpedantic -Wconversion -Wsign-conversion -Wnull-dereference -Wdouble-promotion -Wformat=2 -Wmisleading-indentation -Wduplicated-cond -Wlogical-op -Wuseless-cast -fsanitize=address,thread -fpch-preprocess -v -dD -E
	Using built-in specs.
	COLLECT_GCC=/usr/bin/g++-6
	Target: x86_64-linux-gnu
	Configured with: ../src/configure -v --with-pkgversion='Ubuntu 6.5.0-2ubuntu1~18.04' --with-bugurl=file:///usr/share/doc/gcc-6/README.Bugs --enable-languages=c,ada,c++,go,d,fortran,objc,obj-c++ --prefix=/usr --with-as=/usr/bin/x86_64-linux-gnu-as --with-ld=/usr/bin/x86_64-linux-gnu-ld --program-suffix=-6 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
	Thread model: posix
	gcc version 6.5.0 20181026 (Ubuntu 6.5.0-2ubuntu1~18.04) 
	COLLECT_GCC_OPTIONS='-D' '_GLIBCXX_USE_CXX11_ABI=0' '-D' 'openassetio_EXPORTS' '-g' '-flto' '-fno-fat-lto-objects' '-fvisibility=hidden' '-fvisibility-inlines-hidden' '-Wall' '-Wextra' '-Wshadow' '-Wnon-virtual-dtor' '-Wold-style-cast' '-Wcast-align' '-Wunused' '-Woverloaded-virtual' '-Wpedantic' '-Wconversion' '-Wsign-conversion' '-Wnull-dereference' '-Wdouble-promotion' '-Wformat=2' '-Wmisleading-indentation' '-Wduplicated-cond' '-Wlogical-op' '-Wuseless-cast' '-fsanitize=address,thread' '-fpch-preprocess' '-v' '-dD' '-E' '-D' '___CIDR_DEFINITIONS_END' '-shared-libgcc' '-mtune=generic' '-march=x86-64'
	 /usr/lib/gcc/x86_64-linux-gnu/6/cc1plus -E -quiet -v -imultiarch x86_64-linux-gnu -D_GNU_SOURCE -D _GLIBCXX_USE_CXX11_ABI=0 -D openassetio_EXPORTS -D ___CIDR_DEFINITIONS_END /tmp/compiler-file2857237065531500021 -mtune=generic -march=x86-64 -Wall -Wextra -Wshadow -Wnon-virtual-dtor -Wold-style-cast -Wcast-align -Wunused -Woverloaded-virtual -Wpedantic -Wconversion -Wsign-conversion -Wnull-dereference -Wdouble-promotion -Wformat=2 -Wmisleading-indentation -Wduplicated-cond -Wlogical-op -Wuseless-cast -flto -fno-fat-lto-objects -fvisibility=hidden -fvisibility-inlines-hidden -fsanitize=address,thread -fpch-preprocess -g -fworking-directory -fstack-protector-strong -dD
	cc1plus: error: -fsanitize=address and -fsanitize=kernel-address are incompatible with -fsanitize=thread
	ignoring duplicate directory "/usr/include/x86_64-linux-gnu/c++/6"
	ignoring nonexistent directory "/usr/local/include/x86_64-linux-gnu"
	ignoring nonexistent directory "/usr/lib/gcc/x86_64-linux-gnu/6/../../../../x86_64-linux-gnu/include"
	#include "..." search starts here:
	#include <...> search starts here:
	 /usr/include/c++/6
	 /usr/include/x86_64-linux-gnu/c++/6
	 /usr/include/c++/6/backward
	 /usr/lib/gcc/x86_64-linux-gnu/6/include
	 /usr/local/include
	 /usr/lib/gcc/x86_64-linux-gnu/6/include-fixed
	 /usr/include/x86_64-linux-gnu
	 /usr/include
	End of search list.
```
So I think retaining a more concise manual warning is probably better. We should probably fail the build though, rather than issue a warning and leave one sanitizer enabled and the other disabled.

Reply:
I've changed my mind on this. The CMake disallows simultaneous use of "thread" and "leak" sanitizers, but if I force this then GCC doesn't complain, so I assume they are actually combatible.  The original commit in the source project where I lifted this from (https://github.com/cpp-best-practices/cpp_starter_project/commit/74e8a33a1df351010a4a6f63002e5df93a19c26d) doesn't explain why they guarded it, and I can't find anything on google.

So I conclude it's too error prone and too much maintenance just for somewhat nicer error messages, and I'll remove the CMake checks.  Like I say, GCC fails at CMake configure time if the sanitizer combo is unsupported, so fine.

Reply:
Made CMake error, rather than warn, at configure time in 7ab8022d. But just removed the checks in 7d41159c then, since not much was left, moved out the contents of that .cmake file in f5d89252